Content.Server/GameObjects/Components/Atmos/GasCanisterComponent.cs

```diff
@@ -86,6 +152,187 @@ private void AnchorUpdate()
+            }
+            else if (Air.Pressure < (15 * Atmospherics.OneAtmosphere))
+            {
+	            _appearance?.SetData(GasCanisterVisuals.PressureState, 2);
+            }
+            else
+            {
+	            _appearance?.SetData(GasCanisterVisuals.PressureState, 3);
+            }
+        }
+
+        /// <summary>
+        /// Get the current interface state from server data
+        /// </summary>
+        /// <returns>The state</returns>
+        private GasCanisterBoundUserInterfaceState GetUserInterfaceState()
+        {
+            // We round the pressure for ease of reading
+            return new GasCanisterBoundUserInterfaceState(Label,
+                MathF.Round(Air.Pressure, 2),
+                ReleasePressure,
+                ValveOpened);
+        }
+
+        public void AirWasUpdated()
+        {
+            UpdateUserInterface();
+            UpdateAppearance();
+        }
+
+        #region Check methods
+
+        private bool PlayerCanUse(IEntity? player)
+        {
+            if (player == null)
+            {
+                return false;
+            }
+
+            if (!ActionBlockerSystem.CanInteract(player) ||
+                !ActionBlockerSystem.CanUse(player))
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        #endregion
+
+
+        /// <summary>
+        /// Called when the canister's valve is toggled
+        /// </summary>
+        private void ToggleValve()
+        {
+            ValveOpened = !ValveOpened;
+            UpdateUserInterface();
+        }
+
+        /// <summary>
+        /// Called every frame
+        /// </summary>
+        /// <param name="frameTime"></param>
+        public void Update(in float frameTime)
+        {
+            if (ValveOpened)
+            {
+                var tileAtmosphere = Owner.Transform.Coordinates.GetTileAtmosphere();
+                if (tileAtmosphere != null)
+                {
+                    // Get the GridAtmosphere
+                    var gridId = Owner.Transform.Coordinates.GetGridId(Owner.EntityManager);
+                    var atmosphereSystem = EntitySystem.Get<AtmosphereSystem>();
+                    var gridAtmosphere = atmosphereSystem.GetGridAtmosphere(gridId);
+                    // This shouldn't be null, and is only checked as a failsafe because it could theoretically still be null anyway.
+                    // Releasing gas without invalidation is more dangerous than failing to release.
+                    // Also, if tileAtmosphere.Air is null, then we're airblocked, so DON'T release
+                    if ((gridAtmosphere != null) && (tileAtmosphere.Air != null))
+                    {
+                        Air.ReleaseGasTo(tileAtmosphere.Air, ReleasePressure);
+                        gridAtmosphere.Invalidate(tileAtmosphere.GridIndices);
+                    }
```

Comment:
We have a helper called `InvalidateTileAir` in latest master, that takes in an `EntityCoordinates`. Please use that over doing this manually.

Reply:
I'd need to merge-commit or do a force-push-rebase.